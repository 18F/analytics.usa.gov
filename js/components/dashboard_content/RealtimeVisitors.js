import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";
import formatters from "../../lib/chart_helpers/formatters";

/**
 * Retrieves the realtime users report from the passed data URL and creates a
 * visualization for the count of users visiting sites for the current agency.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @param {string} props.agency the display name for the current agency.
 * @param {number} props.refreshSeconds the number of seconds to wait before
 * refreshing chart data.
 * @returns {import('react').ReactElement} The rendered element
 */
function RealtimeVisitors({ dataHrefBase, agency, refreshSeconds }) {
  const dataURL = `${dataHrefBase}/realtime.json`;
  const ref = useRef(null);
  const [realtimeVisitorData, setRealtimeVisitorData] = useState(null);

  useEffect(() => {
    const initRealtimeVisitorsChart = async () => {
      if (!realtimeVisitorData) {
        const data = await DataLoader.loadRealtimeReportJSON(dataURL);
        await setRealtimeVisitorData(data);
        // Refresh data every interval. useEffect will run and update the chart
        // when the state is changed.
        setInterval(() => {
          DataLoader.loadRealtimeReportJSON(dataURL).then((data) => {
            setRealtimeVisitorData(data);
          });
        }, refreshSeconds * 1000);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder
          .setElement(ref.current)
          .setData(realtimeVisitorData)
          .setRenderer((selection, data) => {
            const totals = data.data[0];
            selection.text(
              totals ? formatters.addCommas(+totals.activeUsers) : 0,
            );
          })
          .build();
      }
    };
    initRealtimeVisitorsChart().catch(console.error);
  }, [realtimeVisitorData]);

  return (
    <div className="text-center tablet:padding-2" ref={ref}>
      <div className="grid-row">
        <span className="data text--bold text--xl grid-col-12">...</span>
      </div>
      <div className="grid-row">
        <div className="text--large desktop:grid-col-8 desktop:grid-offset-2">
          users on {agency} websites and apps in the last 30 minutes
        </div>
      </div>
    </div>
  );
}

RealtimeVisitors.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  agency: PropTypes.string.isRequired,
  refreshSeconds: PropTypes.number.isRequired,
};

export default RealtimeVisitors;
