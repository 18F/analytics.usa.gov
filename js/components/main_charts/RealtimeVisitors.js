import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../lib/chart_helpers/renderblock";
import formatters from "../../lib/chart_helpers/formatters";

/**
 * Retrieves the realtime users report from the passed data URL and creates a
 * visualization for the count of users visiting sites for the current agency.
 *
 * @param {String} dataHrefBase the URL of the base location of the data to be
 * downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @param {String} agency the display name for the current agency.
 */
function RealtimeVisitors({ dataHrefBase, agency }) {
  const reportURL = `${dataHrefBase}/realtime.json`;
  const description = agency ? agency : "U.S. Federal Government";
  const ref = useRef(null);

  useEffect(() => {
    const initRealtimeVisitorsChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: reportURL,
          block: ref.current,
        })
        .call(
          renderBlock.loadAndRender().render((selection, data) => {
            const totals = data.data[0];
            selection.text(formatters.addCommas(+totals.active_visitors));
          }),
        );
      return result;
    };
    initRealtimeVisitorsChart().catch(console.error);
  });

  return (
    <section
      className="chart-realtime__visitors-count"
      data-refresh="15"
      ref={ref}
    >
      <div className="grid-row">
        <h2 className="chart-realtime__current-visitors data grid-col-12">
          ...
        </h2>
      </div>
      <div className="grid-row">
        <div className="chart-realtime__description desktop:grid-col-8 desktop:grid-offset-2">
          people on {description} websites and apps in the last 30 minutes
        </div>
      </div>
    </section>
  );
}

RealtimeVisitors.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  agency: PropTypes.string,
};

export default RealtimeVisitors;
