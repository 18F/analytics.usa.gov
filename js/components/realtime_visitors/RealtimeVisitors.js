import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../lib/chart_helpers/renderblock";
import formatters from "../../lib/chart_helpers/formatters";

function RealtimeVisitors({ dataURL, dataPrefix, agency }) {
  const reportURL = `${dataURL}/${dataPrefix}/realtime.json`;
  const description = agency ? agency : "U.S. Federal Government";
  const ref = useRef(null);

  useEffect(() => {
    const initRealtimeVisitorsChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: reportURL,
          block: ref.current,
        })
        .call(
          renderBlock.loadAndRender().render((selection, data) => {
            const totals = data.data[0];
            selection.text(formatters.addCommas(+totals.active_visitors));
          }),
        );
      return result;
    };
    initRealtimeVisitorsChart().catch(console.error);
  });

  return (
    <section
      className="chart-realtime__visitors-count"
      data-refresh="15"
      ref={ref}
    >
      <div className="grid-row">
        <h2 className="chart-realtime__current-visitors data grid-col-12">
          ...
        </h2>
      </div>
      <div className="grid-row">
        <div className="chart-realtime__description desktop:grid-col-8 desktop:grid-offset-2">
          people on {description} websites and apps in the last 30 minutes
        </div>
      </div>
    </section>
  );
}

RealtimeVisitors.propTypes = {
  dataURL: PropTypes.string.isRequired,
  dataPrefix: PropTypes.string.isRequired,
  agency: PropTypes.string,
};

export default RealtimeVisitors;
