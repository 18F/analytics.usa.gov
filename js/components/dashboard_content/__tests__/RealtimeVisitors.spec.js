import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import RealtimeVisitors from "../RealtimeVisitors";
import DataLoader from "../../../lib/data_loader";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadJSON: jest.fn(),
}));

describe("RealtimeVisitors", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <RealtimeVisitors
          dataHrefBase="http://www.example.com/data/"
          agency="Department of Defense"
          refreshSeconds={30}
        />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    beforeEach(async () => {
      data = {
        name: "realtime",
        query: {
          metrics: [
            {
              name: "activeUsers",
            },
          ],
          samplingLevel: "HIGHER_PRECISION",
          limit: "10000",
          property: "properties/393249053",
        },
        meta: {
          name: "Active Users Right Now",
          description: "Number of users currently visiting all sites.",
        },
        data: [
          {
            active_visitors: "10003538",
          },
        ],
        totals: {},
        taken_at: "2024-01-05T16:05:45.980Z",
      };

      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <RealtimeVisitors
          dataHrefBase="http://www.example.com/data/"
          agency="Department of Defense"
          refreshSeconds={30}
        />,
      );
      await waitFor(() => screen.getByText("10,003,538"));
    });

    it("renders a component with data loaded", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <RealtimeVisitors
          dataHrefBase="http://www.example.com/data/"
          agency="Department of Defense"
          refreshSeconds={30}
        />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("logs the error to console", () => {
      expect(console.error).toHaveBeenCalledWith(error);
    });
  });
});
