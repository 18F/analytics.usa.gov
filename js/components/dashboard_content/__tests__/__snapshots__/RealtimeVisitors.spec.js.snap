// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RealtimeVisitors when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <div
    class="text--centered tablet:padding-2"
  >
    <div
      class="grid-row"
    >
      <span
        class="data text--bold text--xl grid-col-12"
      >
        10,003,538
      </span>
    </div>
    <div
      class="grid-row"
    >
      <div
        class="text--large desktop:grid-col-8 desktop:grid-offset-2"
      >
        users on Department of Defense websites and apps in the last 30 minutes
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RealtimeVisitors when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <div
    class="text--centered tablet:padding-2"
  >
    <div
      class="grid-row"
    >
      <span
        class="data text--bold text--xl grid-col-12"
      >
        ...
      </span>
    </div>
    <div
      class="grid-row"
    >
      <div
        class="text--large desktop:grid-col-8 desktop:grid-offset-2"
      >
        users on Department of Defense websites and apps in the last 30 minutes
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RealtimeVisitors when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <div
    class="text--centered tablet:padding-2"
  >
    <div
      class="grid-row"
    >
      <span
        class="data text--bold text--xl grid-col-12"
      >
        ...
      </span>
    </div>
    <div
      class="grid-row"
    >
      <div
        class="text--large desktop:grid-col-8 desktop:grid-offset-2"
      >
        users on Department of Defense websites and apps in the last 30 minutes
      </div>
    </div>
  </div>
</DocumentFragment>
`;
